File: src/backend/replication/biz-replication/src/main/java/com/tencent/secapi/checkurl/ValidateUrlFunction.java
package com.tencent.secapi.checkurl;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@FunctionalInterface
public interface ValidateUrlFunction {

    Map<String, Pattern> PATTERN_CACHE = new ConcurrentHashMap<>();

    boolean validateUrl(List<String> rules, String domain);

    ValidateUrlFunction IS_DOMAIN = (domains, equalDomain) -> domains.contains(equalDomain);

    ValidateUrlFunction IS_SUB_DOMAIN = (domains, subdomain) -> {
        for (String domain : domains) {
            if (!domain.trim().isEmpty()) {
                if (subdomain.equals(domain) || subdomain.endsWith("." + domain)) {
                    return true;
                }
            }
        }
        return false;
    };

    ValidateUrlFunction IS_SUB_DOMAIN_REGEX = (domainRegexs, domain) -> {
        for (String regex : domainRegexs) {
            Pattern pattern = PATTERN_CACHE.computeIfAbsent(regex, Pattern::compile);
            Matcher matcher = pattern.matcher(domain);
            if (matcher.find()) {
                return true;
            }
        }
        return false;
    };
}

File: src/backend/common/common-api/src/main/kotlin/com/tencent/bkrepo/common/api/util/checkurl/ValidateUrlFunction.kt
package com.tencent.bkrepo.common.api.util.checkurl

import java.util.concurrent.ConcurrentHashMap

fun interface ValidateUrlFunction {
    fun validateUrl(rules: List<String>, domain: String): Boolean

    companion object {
        private val patternCache = ConcurrentHashMap<String, Regex>()

        val IS_DOMAIN = ValidateUrlFunction { rules, domain -> rules.contains(domain) }

        val IS_SUB_DOMAIN = ValidateUrlFunction { rules, subdomain ->
            rules.any { domain ->
                domain.isNotBlank() && (subdomain == domain || subdomain.endsWith(".$domain"))
            }
        }

        val IS_SUB_DOMAIN_REGEX = ValidateUrlFunction { domainRegexes, domain ->
            domainRegexes.any { regex ->
                val pattern = patternCache.computeIfAbsent(regex, ::Regex)
                pattern.containsMatchIn(domain)
            }
        }
    }
}
